Emby.Server.Implementations/LiveTv/Listings/XmlTvListingsProvider.cs

```diff
@@ -67,110 +64,65 @@ private async Task<string> GetXml(ListingsProviderInfo info, CancellationToken c
-            {
-                return UnzipIfNeeded(info.Path, info.Path);
-            }
-
             string cacheFilename = info.Id + ".xml";
             string cacheFile = Path.Combine(_config.ApplicationPaths.CachePath, "xmltv", cacheFilename);
+
             if (File.Exists(cacheFile) && File.GetLastWriteTimeUtc(cacheFile) >= DateTime.UtcNow.Subtract(_maxCacheAge))
             {
-                return UnzipIfNeeded(info.Path, cacheFile);
+                return cacheFile;
             }
 
             // Must check if file exists as parent directory may not exist.
             if (File.Exists(cacheFile))
             {
                 File.Delete(cacheFile);
             }
+            else
+            {
+                Directory.CreateDirectory(Path.GetDirectoryName(cacheFile));
+            }
 
-            _logger.LogInformation("Downloading xmltv listings from {Path}", info.Path);
+            if (info.Path.StartsWith("http", StringComparison.OrdinalIgnoreCase))
+            {
+                _logger.LogInformation("Downloading xmltv listings from {Path}", info.Path);
 
-            Directory.CreateDirectory(Path.GetDirectoryName(cacheFile));
+                using var response = await _httpClientFactory.CreateClient(NamedClient.Default).GetAsync(info.Path, cancellationToken).ConfigureAwait(false);
+                await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
 
-            using var response = await _httpClientFactory.CreateClient(NamedClient.Default).GetAsync(info.Path, cancellationToken).ConfigureAwait(false);
-            await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
-            await using (var fileStream = new FileStream(cacheFile, FileMode.CreateNew, FileAccess.Write, FileShare.None, IODefaults.CopyToBufferSize, FileOptions.Asynchronous))
-            {
-                await stream.CopyToAsync(fileStream, cancellationToken).ConfigureAwait(false);
+                return await UnzipIfNeededAndCopy(info.Path, stream, cacheFile, cancellationToken).ConfigureAwait(false);
             }
+            else
+            {
+                await using var stream = new FileStream(info.Path, FileMode.Open, FileAccess.Read, FileShare.Read, IODefaults.FileStreamBufferSize, FileOptions.Asynchronous);
 
-            return UnzipIfNeeded(info.Path, cacheFile);
+                return await UnzipIfNeededAndCopy(info.Path, stream, cacheFile, cancellationToken).ConfigureAwait(false);
+            }
         }
 
-        private string UnzipIfNeeded(ReadOnlySpan<char> originalUrl, string file)
+        private async Task<string> UnzipIfNeededAndCopy(string originalUrl, Stream stream, string file, CancellationToken cancellationToken)
         {
-            ReadOnlySpan<char> ext = Path.GetExtension(originalUrl.LeftPart('?'));
+            string ext = Path.GetExtension(originalUrl.Split('?')[0]);
```

Comment:
Why change this?

Reply:
Can't use spans in async methods, but Split instead of substring with indexof isn't ideal

Reply:
I used Split() for readability. Should it be something like this?
```
int index = originalUrl.IndexOf('?', StringComparison.CurrentCulture);
string ext = Path.GetExtension(index > -1 ? originalUrl.Remove(index) : originalUrl);
```